### PR TITLE
Implement Add Gem 

### PR DESCRIPTION
**Pull request by 3umarG**
#### Add gem endpoint including some edge cases : 
- Add gem as _command_, with ability to add without links or description "easy as possible".
- Handling repeated gems with the same link for the same user.
- Handling not repeatable tags.
- Unique formatted tags names for all users.
- Special tags with lables for each user "for preview"

## Changes checklist

- [X] I tested my solution locally.
- [X] I checked if my solution impacts the checklist/requirements.
- [X] I wrote _integration_ tests.

### CODE REVIEW

File: .jpb/jpb-settings.xml
Comment (xaaksw) [inline .jpb/jpb-settings.xml:1]: I think this file should be in `.gitignore` right?

File: src/main/kotlin/com/re/back/gems/dtos/request/GemRequestDto.kt [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+import jakarta.validation.constraints.*
+
+data class GemRequestDto(
+    @field:NotBlank val title: String,
+    val description: String? = null,
+    val link: String? = null,
+    val isPublic: Boolean = false,
+    val isOriginalContent: Boolean = false,
+    @field:NotNull val isCommand: Boolean,
```
Comment (xaaksw) [inline src/main/kotlin/com/re/back/gems/dtos/request/GemRequestDto.kt]: `Command` is one of our tags , not for every gem

File: src/main/kotlin/com/re/back/gems/dtos/response/GemResponseDto.kt [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+data class GemResponseDto(
+    val id: Int,
+    val title: String,
+    val description: String? = null,
+    val link: String? = null,
+    val isPublic: Boolean,
+    val isOriginalContent: Boolean,
+    val createdOn: LocalDateTime,
+    val updatedOn: LocalDateTime?,
+    val isCommand: Boolean,
```
Comment (xaaksw) [inline src/main/kotlin/com/re/back/gems/dtos/response/GemResponseDto.kt]: same comment here

File: src/main/kotlin/com/re/back/gems/entities/Gem.kt [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    val isOriginalContent: Boolean = false,
+    @Column(nullable = false) val createdOn: LocalDateTime = LocalDateTime.now(),
+    @Column(nullable = true) val updatedOn: LocalDateTime? = null,
+
+    @ManyToOne(fetch = FetchType.LAZY, optional = false)
+    @JoinColumn(
+        name = "user_id",
+        nullable = false
+    ) val user: AppUser,
+    @Column(nullable = false) val isCommand: Boolean = false,
```
Comment (xaaksw) [inline src/main/kotlin/com/re/back/gems/entities/Gem.kt]: same comment here

File: src/main/kotlin/com/re/back/gems/services/GemsService.kt [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+                val gemWithSameLink = gemsRepository.findByUserIdAndLink(user.id!!, link)
+                if (gemWithSameLink.isPresent) {
+                    throw AlreadyUsedLinkException(gemWithSameLink.get().toGemResponseDto())
+                }
+            }
+        }
+    }
+
+    private fun checkForGemRequestDto(gemRequestDto: GemRequestDto) {
+        if (!gemRequestDto.isCommand && (gemRequestDto.link.isNullOrEmpty() || gemRequestDto.link.isBlank())) {
```
Comment (xaaksw) [inline src/main/kotlin/com/re/back/gems/services/GemsService.kt]: 'isBlank' check is enough here

File: src/test/kotlin/com/re/back/integration_tests/gems/GemsControllerTests.kt [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+        assertEquals((responseEntity.body?.data as LinkedHashMap<*, *>)["link"], link)
+
+    }
+
+    @Test
+    fun `Add gem without any tags, return 200 OK`() {
+        // Arrange
+        val title = "gem 1"
+        val link = "link1"
+        val isCommand = false
```
Comment (xaaksw) [inline src/test/kotlin/com/re/back/integration_tests/gems/GemsControllerTests.kt]: Same comment for dealing with command as a tag

File: src/main/kotlin/com/re/back/gems/services/GemsService.kt (hunk 4 above)
Comment (3umarG) [reply src/main/kotlin/com/re/back/gems/services/GemsService.kt]: `isBlank` is available only on `CharSequence` not in nullable `CharSequence?`, you must prefix it with nullable check

**Review by xaaksw — APPROVED** [review]

**Pull request merged**